File: typescript/tracing/langchain/27_tool_error.ts
import { tool } from "langchain";
import { z } from "zod";

import { initRespan, shutdown, tracingConfig } from "./_shared";

const failingTool = tool(
  ({ key }: { key: string }) => {
    throw new Error(`no result for ${key}`);
  },
  {
    name: "failing_lookup",
    description: "Fail for missing keys.",
    schema: z.object({ key: z.string() }),
  },
);

export async function toolError(): Promise<void> {
  const runtime = await initRespan("typescript-langchain-tool-error");

  try {
    await failingTool.invoke(
      { key: "missing" },
      tracingConfig(runtime, "tool_error"),
    );
  } catch (error) {
    console.log(`caught expected error: ${(error as Error).message}`);
  } finally {
    await shutdown(runtime);
  }
}

if (import.meta.url === `file://${process.argv[1]}`) {
  await toolError();
}

File: python/tracing/langchain/27_tool_error.py
"""Tool error."""

from langchain_core.tools import tool

from _shared import flush, init_telemetry, tracing_config


@tool
def failing_lookup(query: str) -> str:
    """Always fail to demonstrate tool error callbacks."""
    raise ValueError(f"no result for {query}")


def tool_error() -> None:
    telemetry = init_telemetry("langchain-tool-error")
    try:
        try:
            failing_lookup.invoke(
                {"query": "missing"},
                config=tracing_config("tool_error"),
            )
        except ValueError as exc:
            print(f"caught: {exc}")
    finally:
        flush(telemetry)


if __name__ == "__main__":
    tool_error()
